src/v/datalake/arrow_schema_translator.cc

```diff
@@ -196,24 +251,297 @@ struct schema_converter_builder_visitor {
+    // values that's probably ok, but this will cause problems with seastar
+    // allocation limits. If we're going to continue using Arrow we should
+    // make an ArrayBuilder for iobuf and/or ss:sstring.
+    const auto& val = std::get<iceberg::string_value>(prim_value).val;
+    auto begin = iobuf::byte_iterator(val.cbegin(), val.cend());
+    auto end = iobuf::byte_iterator(val.cend(), val.cend());
+    std::string str;
+    std::copy(begin, end, std::back_inserter(str));
+    auto status = _builder->Append(str);
+    if (!status.ok()) {
+        throw std::runtime_error("Unable to add value to string builder");
+    }
+}
+
+template<>
+void scalar_schema_converter<arrow::FixedSizeBinaryType>::add_primitive(
+  const iceberg::primitive_value& prim_value) {
+    if (std::holds_alternative<iceberg::fixed_value>(prim_value)) {
+        // TODO: This copies the entire string into a std::string. For small
+        // values
+        // that's probably ok, but this will cause problems with seastar
+        // allocation limits. If we're going to continue using Arrow we should
+        // make an ArrayBuilder for iobuf and/or ss:sstring.
+        const auto& val = std::get<iceberg::fixed_value>(prim_value).val;
+        auto begin = iobuf::byte_iterator(val.cbegin(), val.cend());
+        auto end = iobuf::byte_iterator(val.cend(), val.cend());
+        std::string str;
+        std::copy(begin, end, std::back_inserter(str));
+        if (str.size() != static_cast<size_t>(_builder->byte_width())) {
+            throw std::invalid_argument(fmt::format(
+              "Expected {} byte value. Got {}.",
+              static_cast<size_t>(_builder->byte_width()),
+              str.size()));
+        }
+        auto status = _builder->Append(str);
+        if (!status.ok()) {
+            throw std::runtime_error(
+              "Unable to add value to fixed-size binary builder");
+        }
+    } else if (std::holds_alternative<iceberg::uuid_value>(prim_value)) {
+        const auto& val
+          = std::get<iceberg::uuid_value>(prim_value).val.to_vector();
+        auto status = _builder->Append(val.data());
+        if (!status.ok()) {
+            throw std::runtime_error(
+              "Unable to add value to fixed-size binary builder");
+        }
+    } else {
+        throw std::invalid_argument(
+          "Scalar converter expected a fixed or uuid");
+    }
+}
+
+template<>
+void scalar_schema_converter<arrow::BinaryType>::add_primitive(
+  const iceberg::primitive_value& prim_value) {
+    if (std::holds_alternative<iceberg::binary_value>(prim_value)) {
+        const auto& val = std::get<iceberg::binary_value>(prim_value).val;
+        auto begin = iobuf::byte_iterator(val.cbegin(), val.cend());
+        auto end = iobuf::byte_iterator(val.cend(), val.cend());
+        std::string str;
+        std::copy(begin, end, std::back_inserter(str));
+        auto status = _builder->Append(str);
```

Comment:
It looks like BaseBinaryBuilder has an Append(char*, size_t) interface. If that's available then it seems like you can append from the iobuf fragments directly into the builder without building the intermediate string.